apps/src/blockly/eventHandlers.ts

```diff
@@ -117,3 +122,78 @@ export function reflowToolbox() {
     modalWorkspace?.getFlyout()?.reflow();
   }
 }
+export function updateBlockLimits(event: Abstract) {
+  // This check is to update bubbles that show block limits whenever
+  // blocks on the main workspace are updated.
+  if (
+    event.type !== Blockly.Events.BLOCK_CHANGE &&
+    event.type !== Blockly.Events.BLOCK_MOVE &&
+    event.type !== Blockly.Events.BLOCK_CREATE &&
+    event.type !== Blockly.Events.THEME_CHANGE
+  ) {
+    return;
+  }
+  const blockEvent = event as
+    | BlockChange
+    | BlockMove
+    | BlockCreate
+    | ThemeChange;
+  const blockLimitMap = Blockly.blockLimitMap;
+
+  if (!blockEvent.workspaceId || !blockLimitMap || !(blockLimitMap?.size > 0)) {
+    return;
+  }
+  const eventWorkspace = Blockly.Workspace.getById(
+    blockEvent.workspaceId
+  ) as ExtendedWorkspaceSvg;
+  const allWorkspaceBlocks = eventWorkspace?.getAllBlocks();
+  if (!allWorkspaceBlocks) {
+    return;
+  }
+  // Define a Map to store block counts for each type
+  const blockCountMap = new Map<string, number>();
+  Blockly.blockCountMap = blockCountMap;
+  // Initialize block counts based on blockLimitMap
+  blockLimitMap.forEach((_, type) => {
+    blockCountMap.set(type, 0);
+  });
+
+  // Count the enabled blocks of each type
+  allWorkspaceBlocks
+    .filter(block => blockLimitMap.has(block.type) && block.isEnabled())
+    .forEach(block => {
+      const type = block.type;
+      if (blockCountMap.has(type)) {
+        const currentCount = blockCountMap.get(type) || 0;
+        blockCountMap.set(type, currentCount + 1);
+      }
+    });
+
+  const flyout = eventWorkspace.getFlyout();
+  if (!flyout) {
+    return;
+  }
+  // Get all blocks from the flyout
+  const flyoutBlocks = flyout
+    .getWorkspace()
+    .getTopBlocks() as ExtendedBlockSvg[];
+
+  // Get the flyout blocks that have limits
+  const limitedFlyoutBlocks = flyoutBlocks.filter(block => {
+    return blockLimitMap.has(block.type);
+  });
+
+  limitedFlyoutBlocks.forEach(flyoutBlock => {
```

Comment:
nit: If you're going to do a filter and then a forEach, I have a slight preference for just doing a forEach with an early return,

```
flyoutBlocks.forEach(block => {
    const hasLimit = blockLimitMap.has(block.type);
    if (!hasLimit) {
      return;
    }
   ...
}
```
Since you retrieve the limit later anyway, you might be able to do...
```
flyoutBlocks.forEach(block => {
    const limit = blockLimitMap.get(block.type);
    if (!limit) { // depending on how to handle zero
      return;
    }
    const blockUsedCount = blockCountMap.get(flyoutBlock.type) || 0;
   ...
}
```